### DISCUSSION TITLE
The few ideas of how I envision such a system

### DISCUSSION THREAD

**Discussion by En3Tho** [ScottArbeit/Grace#14] [Ideas]
First of all, thank you for the presentation. It was really thoughtful and nice. Additional like for and F# choice (alrhought you have to be careful with it's performance as most construct allocate and compiler really likes to put tailcall prefixes everywehre which jit doesn't like at all e.t.c.).

What I envision in the first place is not just a version control system. Such a system can be turned into something more. A developer community right here in your editor. You could:

1) See what people are working on, join in, talk about stuff.
2) Ask questions, get reviews right there
3) Get notifications about review changes, instantly switch branches with auto-save, no manual repetitive stuff needed
4) Such conversations about code could be used by AI to generate a comprehensive documentation and also be checked if code is right.
5) Obviously, auto maintaining of forks (well, branches) is a super cool stuff
6) Gettting automatic help from AI based on documentation or some other source of data

Usually all these things are done separately now. Some conversations in GitHub, some in Discord, some in private. But what if relevant stuff could be discussed right there with the code itself? I kinda see this as more than just "well, keep track of these files" kind of thing but instead "use all the information about files and changes and be of real help right here and now".

P.S. I've even written about this in youtube comments section (which I rarely do) because I was really surprised at people's reaction to this. It's almost like they don't use internet (commenting on youtube might be the only rare exception it seems).

**Comment by ScottArbeit** [comment]
I do think of Grace as being a platform to build features on. There have been attempts over the years to use Git as a database to enable some subset of what you're proposing, and none of them have ever really caught on... mostly because Git is not a database. 🤦‍♂️

There's a line to walk between doing features that replace Git, and features that replace GitHub / GitLab / BitBucket / etc. Given how early it is, I'm trying to stay mostly on the side of version control itself, and not platform features on top of it, but I definitely want to make sure that the eventing model of Grace supports all kinds of new features, some of which you can imagine as the things that GitHub already does to help with code (Copilot, secret scanning, security vulnerability scanning, etc.), and most of which will be suggested by you and other users. I have a few ideas, but there's no way I have all or most of them.

With that said, one area that I think it's valuable and important to build into Grace itself is some interface with LLM's. One example I mentioned in my talk is: if a promotion conflict is detected, don't just tell the user about it, but also submit the code to an LLM and get recommendations about how to handle the conflict, and then give all of that context to the user in one place. The understandable fears about auto-rebasing hopefully go away when people start to see it working really well 95% of the time, and getting detailed help when user intervention is required.

**Reply by En3Tho** [reply]
Thank you for the detailed response. I just thought that what you're doing is really cool and I personally feel like this has a lot of potential.

I understand that first of all this should be a vcs and it should be good at what it is intended to do. And you right to think about it from the vcs point of view.

It's just that your talk made me think again of what difficulties I experience as a developer. Not only git related but overall. Building things, maintaining code, finding information. Thank you for this kind of think-push.

I really hope this eventing model will be able to power much more, a true developer space not requiring to chat in one place, file issues in another, discuss and run snippets in third etc. These are not strictly vcs features but I believe for them to work seamlessly a first class support from vcs is a must.
